README
NAME
    Task::Dancer - Dancer in a box

VERSION
    Version 0.38

DESCRIPTION
    This Task installs Dancer, optional engines, templates and assorted
    modules that are not included in the Dancer core distribution.

    If you've written anything relating to Dancer, please let us know.

    We try to maintain a list of modules that are maintained and install
    properly. If any of the modules "Task::Dancer" tried to install failing,
    warn us, so we can temporarily disable it. Also, tell us if any module
    temporarily disabled is working again.

INCLUDES MODULES
  Template Engines
    Dancer::Template::Alloy
    Dancer::Template::Haml
    Dancer::Template::HtmlTemplate
    Dancer::Template::Mason
    Dancer::Template::Mason2
    Dancer::Template::MicroTemplate
    Dancer::Template::MojoTemplate
    Dancer::Template::Semantic (disabled)
    Dancer::Template::TemplateFlute
    Dancer::Template::TemplateSandbox
    Dancer::Template::Tenjin
    Dancer::Template::Tiny
    Dancer::Template::Xslate

  Logging Engines
    Dancer::Logger::ColorConsole
    Dancer::Logger::ConsoleSpinner
    Dancer::Logger::Log4perl
    Dancer::Logger::LogHandler
    Dancer::Logger::Pipe
    Dancer::Logger::PSGI
    Dancer::Logger::Syslog

  Serialization
    Dancer::Serializer::UUEncode

  Session Engines
    Dancer::Session::CHI
    Dancer::Session::Cookie
    Dancer::Session::KiokuDB
    Dancer::Session::Memcached
    Dancer::Session::MongoDB
    Dancer::Session::PSGI
    Dancer::Session::Storable

  Plugins
    Dancer::Plugin::Async
    Dancer::Plugin::Auth::Basic
    Dancer::Plugin::Auth::Htpasswd
    Dancer::Plugin::Auth::RBAC
    Dancer::Plugin::Auth::Twitter
    Dancer::Plugin::Bcrypt
    Dancer::Plugin::Browser
    Dancer::Plugin::Cache::CHI
    Dancer::Plugin::Captcha::SecurityImage
    Dancer::Plugin::Database
    Dancer::Plugin::DBIC
    Dancer::Plugin::DebugDump
    Dancer::Plugin::DebugToolbar
    Dancer::Plugin::DirectoryView
    Dancer::Plugin::Dispatcher
    Dancer::Plugin::ElasticSearch
    Dancer::Plugin::Email
    Dancer::Plugin::EncodeID
    Dancer::Plugin::EscapeHTML
    Dancer::Plugin::Facebook
    Dancer::Plugin::Fake::Response
    Dancer::Plugin::Feed
    Dancer::Plugin::FlashMessage
    Dancer::Plugin::FlashNote
    Dancer::Plugin::FormattedOutput
    Dancer::Plugin::FormValidator
    Dancer::Plugin::Hosts
    Dancer::Plugin::LibraryThing
    Dancer::Plugin::Memcached
    Dancer::Plugin::MemcachedFast
    Dancer::Plugin::MobileDevice
    Dancer::Plugin::Mongo
    Dancer::Plugin::Mongoose
    Dancer::Plugin::MPD
    Dancer::Plugin::Nitesi
    Dancer::Plugin::NYTProf
    Dancer::Plugin::ORMesque
    Dancer::Plugin::Params::Normalization
    Dancer::Plugin::Passphrase
    Dancer::Plugin::Preprocess::Sass
    Dancer::Plugin::Progress
    Dancer::Plugin::ProxyPath
    Dancer::Plugin::Redis
    Dancer::Plugin::REST
    Dancer::Plugin::Scoped (temporarily disabled)
    Dancer::Plugin::SimpleCRUD
    Dancer::Plugin::SiteMap
    Dancer::Plugin::SMS
    Dancer::Plugin::SporeDefinitionControl (disabled)
    Dancer::Plugin::Stomp
    Dancer::Plugin::TimeRequest
    Dancer::Plugin::Thumbnail
    Dancer::Plugin::ValidateTiny
    Dancer::Plugin::ValidationClass
    Dancer::Plugin::WebSocket
    Dancer::Plugin::XML::RSS

More Plack middlewares
    Dancer::Middleware::Rebase
    Dancer::Debug

AUTHOR
    Sawyer X, "xsawyerx AT cpan DOT org"

    Alberto Simoes, "ambs AT cpan DOT org"

BUGS
    Please report any bugs or feature requests to "bug-task-dancer at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Task-Dancer>. I will be
    notified, and then you'll automatically be notified of progress on your
    bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Task::Dancer

    You can also look for information at:

    *   RT: CPAN's request tracker (report bugs here)

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Task-Dancer>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Task-Dancer>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Task-Dancer>

    *   Search CPAN

        <http://search.cpan.org/dist/Task-Dancer/>

ACKNOWLEDGEMENTS
    Dancer team.

LICENSE AND COPYRIGHT
    Copyright 2010-2012 Sawyer X.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either: the GNU General Public License as published
    by the Free Software Foundation; or the Artistic License.

    See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for Task-Dancer

0.38    24.12.12
        * No functional changes.

0.37    23.10.12
        * No functional changes. Released using Module::Install 1.06.

0.36    06.02.12
        * Fixing Dancer::Plugin::Haml spelling (Brian Hann - c0bra).

0.35    13.11.11
        * Adding Dancer::Plugin::Auth::Basic
        * Adding Dancer::Plugin::Auth::Htpasswd
        * Adding Dancer::Plugin::Nitesi
        * Enabling Dancer::Session::CHI
        * Removing Dancer::Plugin::DataFu
        * Forcing a specific version for Dancer::Plugin::NYTProf
        * Check Perl versions before installing
          (for now, useful to skip Mojolicious)

Failing:
 - Dancer::Plugin::ValidateTiny  (perl version, kalgan)
 - Dancer::Plugin::Params::Normalization (undef obj call method, dams)
 - Dancer::Logger::PSGI (tests, franck)

0.34    05.11.11
        * Adding Dancer::Plugin::Preprocess::Sass
        * Adding Dancer::Plugin::Hosts
        * Adding Dancer::Plugin::Passphrase
        * Adding Dancer::Plugin::ValidateTiny
        * Disabling Dancer::Plugin::SporeDefinitionControl
        * Disabling Dancer::Template::Semantic
        * Adding and disabling Dancer::Session::CHI

0.33    21.09.11
        * Adding Dancer::Plugin::Dispatcher
        * Adding Dancer::Plugin::Mongoose
        * Adding Dancer::Plugin::EncodeID

0.32    08.09.11
        * Adding Dancer::Plugin::ElasticSearch
        * Adding Dancer::Plugin::DirectoryView
        * Adding Dancer::Plugin::Stomp
        * Adding Dancer::Plugin::LibraryThing

0.31    11.08.11 
        * Adding Dancer::Plugin::DebugToolbar
        * Adding Dancer::Plugin::EscapeHTML
        * Uncommenting Dancer::Plugin::ORMesque
        * Removing deprecated Dancer::Plugin::Validation         

0.30    30.07.11
        * Adding Dancer::Plugin::NYTProf
        * Adding Dancer::Plugin::Thumbnail

0.29    03.07.11
        * Adding Dancer::Plugin::Bcrypt
        * Adding Dancer::Plugin::XML::RSS
        * Adding Dancer::Session::MongoDB
        * Adding Dancer::Template::TemplateFlute
        * Adding Dancer::Logger::ColorConsole
        * Uncommenting Dancer::Plugin::Mongo

0.28    28.05.11
        * Commenting Dancer::Plugin::Mongo (MongoDB not installing)
        * Adding Dancer::Plugin::TimeRequests
        * Adding Dancer::Plugin::Async

0.27    11.04.11
        * Adding Dancer::Plugin::Fake::Response
        * Deprecating Dancer::Plugin::Cache, and
        * Adding Dancer::Plugin::Cache::CHI
        * Adding Dancer::Plugin::Progress
        * Uncommenting Dancer::Plugin::MobileDevice (forcing 0.03)

0.26    29.03.11
        * Fixed Makefile.PL and behavior with default modules
        * Adding Dancer::Plugin::Captcha::SecurityImage
        * Uncommenting Dancer::Plugin::ProxyPath (working with latest Dancer)

0.25    20.03.11
        * Adding Dancer::Plugin::Cache

0.24    11.03.11
        * Adding Dancer::Logger::Log4perl
        * Uncommenting Dancer::Session::KiokuDB (forcing 0.04)
        * Uncommenting Dancer::Template::Mason2 (forcing 0.02)

0.23    07.03.11
        * Adding Dancer::Logger::Pipe
        * Adding Dancer::Template::Xslate
        * Adding Dancer::Template::MojoTemplate
        * Adding Dancer::Middleware::Reabase
        * Adding Dancer::Template::Semantic
        * Adding Dancer::Session::KiokuDB
        * Adding Dancer::Serializer::UUEncode
        * Adding Dancer::Logger::ConsoleSpinner
        * Forcing Dancer::Plugin::SimpleCRUD, version 0.02 (for Dancer 1.3)
        * Uncommenting Dancer::Plugin::SporeDefinitionControl
        * Adding and commenting Dancer::Template::Mason2 (failing 0.01)
        * Removing Dancer::Plugin::Authorize (deprecated)

0.22    05.03.11
        * Adding Dancer::Plugin::Auth::RBAC
        * Adding Dancer::Plugin::Auth::Twitter
        * Adding Dancer::Plugin::DebugDump
        * Adding Dancer::Plugin::MemcachedFast
        * Adding Dancer::Plugin::Facebook
        * Adding Dancer::Plugin::FormattedOutput
        * Uncommenting Dancer::Plugin::Rest (force 0.05)
        * Adding and commenting Dancer::Plugin::GearmanXS (failing 0.110570)
        * Adding and commenting Dancer::Plugin::ProxyPath (failing 0.01)

0.21    16.02.11
        * Adding Dancer::Plugin::DataFu
        * Adding Dancer::Plugin::Feed
        * Adding Dancer::Plugin::FlashNote
        * Adding Dancer::Plugin::FlashMessage
        * Adding Dancer::Plugin::FormValidator
        * Adding Dancer::Plugin::Memcached
        * Adding Dancer::Plugin::MPD
        * Adding Dancer::Plugin::Redis
        * Adding Dancer::Plugin::SMS
        * Adding Dancer::Plugin::WebSocket
        * Adding Dancer::Plugin::SimpleCRUD
        * Adding Dancer::Plugin::ValidationClass
        * Adding and commenting Dancer::Plugin::MobileDevice
        * Adding and commenting Dancer::Plugin::SporeDefinitionControl
        * Removing Dancer::Template::Ctpp2 (not automatic)
        * Commenting Dancer::Plugin::ORMesque (failing 1.103180)
        * Commenting Dancer::Plugin::Validation (not on CPAN?)
        * Commenting Dancer::Plugin::Scoped (failing 0.01)
        * Commenting Dancer::Plugin::REST (failing 0.04)

0.20    02.11.10
        * Adding Dancer::Plugin::ORMesque
        * Adding Dancer::Plugin::Browser

0.19    28.10.10
        * Adding Dancer::Plugin::Mongo

0.18    20.10.10
        * Fixed typo for Dancer::Plugin::Scoped
        (thanks to nmani - Naveen)

0.17    12.10.10
        * Added Dancer::Plugin::Params::Normalization

0.16    14.09.10
        * Install everything by default
        * Added Dancer::Plugin::Authorize

0.15    12.09.10
        * Fixed POD
        * Fixed Changes

0.14    11.09.10
        * Adding Dancer::Plugin::DBIC

0.13    12.07.10
        * Adding Dancer::Logger::PSGI
        * Adding Dancer::Session::PSGI
        * Adding Dancer::Plugin::Scoped
        * Adding Dancer::Debug

0.12    24.06.10
        * Adding Dancer::Template::Ctpp2

0.11    28.05.10
        * Adding Dancer::Template::Alloy
        * Adding Dancer::Plugin::REST

Build.PL
use 5.006;
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'Task::Dancer',
    license             => 'perl',
    dist_author         => q{Sawyer X <xsawyerx@cpan.org>},
    dist_version_from   => 'lib/Task/Dancer.pm',
    build_requires => {
        'Test::More' => 0,
    },

    requires => {
        'perl' => 5.006,
    },

    auto_features => {
        template_tiny => {
            description => 'Template::Tiny templating support',
            requires    => { 'Dancer::Template::Tiny' => 0 },
        },

        template_tenjin => {
            description => 'Tenjin templating support',
            requires    => {
                'Dancer::Template::Tenjin' => 0,
                'Tenjin'                   => 0.5,
            },
        },

        template_haml => {
            description => 'Text::HAML templating support',
            requires    => { 'Dancer::Template::HAML' => 0 },
        },

        template_microtemplate => {
            description => 'Text::MicroTemplate templating support',
            requires    => { 'Dancer::Template::MicroTemplate' => 0 },
        },

        logging_syslog => {
            description => 'Sys::Syslog logging support',
            requires    => { 'Dancer::Logger::Syslog' => 0 },
        },

        logging_loghandler => {
            description => 'Log::Handler logging support',
            requires    => { 'Dancer::Logger::LogHandler' => 0 },
        },

        session_memcache => {
            description => 'Cache::Memcached session support',
            requires    => { 'Dancer::Session::Memcache' => 0 },
        },

        session_cookie => {
            description => 'Encrypted cookie-based session support',
            requires    => { 'Dancer::Session::Cookie' => 0 },
        },

        session_storable => {
            description => 'Storable file-based session support',
            requires    => { 'Dancer::Session::Storable' => 0 },
        },
    },

    add_to_cleanup      => [ 'Task-Dancer-*' ],
);

$builder->create_build_script();


lib/Task/Dancer.pm
package Task::Dancer;
use strict;
use warnings;
our $VERSION = '0.38';

1;

__END__

=head1 NAME

Task::Dancer - Dancer in a box

=head1 VERSION

Version 0.38

=head1 DESCRIPTION

This Task installs Dancer, optional engines, templates and assorted
modules that are not included in the Dancer core distribution.

If you've written anything relating to Dancer, please let us know.

We try to maintain a list of modules that are maintained and install
properly. If any of the modules C<Task::Dancer> tried to install
failing, warn us, so we can temporarily disable it. Also, tell us
if any module temporarily disabled is working again.

=head1 INCLUDES MODULES

=head2 Template Engines

=over 4

=item L<Dancer::Template::Alloy>

=item L<Dancer::Template::Haml>

=item L<Dancer::Template::HtmlTemplate>

=item L<Dancer::Template::Mason>

=item L<Dancer::Template::Mason2>

=item L<Dancer::Template::MicroTemplate>

=item L<Dancer::Template::MojoTemplate>

=item L<Dancer::Template::Semantic> (disabled)

=item L<Dancer::Template::TemplateFlute>

=item L<Dancer::Template::TemplateSandbox>

=item L<Dancer::Template::Tenjin>

=item L<Dancer::Template::Tiny>

=item L<Dancer::Template::Xslate>

=back

=head2 Logging Engines

=over 4

=item L<Dancer::Logger::ColorConsole>

=item L<Dancer::Logger::ConsoleSpinner>

=item L<Dancer::Logger::Log4perl>

=item L<Dancer::Logger::LogHandler>

=item L<Dancer::Logger::Pipe>

=item L<Dancer::Logger::PSGI>

=item L<Dancer::Logger::Syslog>

=back

=head2 Serialization

=over 4

=item L<Dancer::Serializer::UUEncode>

=back

=head2 Session Engines

=over 4

=item L<Dancer::Session::CHI>

=item L<Dancer::Session::Cookie>

=item L<Dancer::Session::KiokuDB>

=item L<Dancer::Session::Memcached>

=item L<Dancer::Session::MongoDB>

=item L<Dancer::Session::PSGI>

=item L<Dancer::Session::Storable>

=back

=head2 Plugins

=over 4

=item L<Dancer::Plugin::Async>

=item L<Dancer::Plugin::Auth::Basic>

=item L<Dancer::Plugin::Auth::Htpasswd>

=item L<Dancer::Plugin::Auth::RBAC>

=item L<Dancer::Plugin::Auth::Twitter>

=item L<Dancer::Plugin::Bcrypt>

=item L<Dancer::Plugin::Browser>

=item L<Dancer::Plugin::Cache::CHI>

=item L<Dancer::Plugin::Captcha::SecurityImage>

=item L<Dancer::Plugin::Database>

=item L<Dancer::Plugin::DBIC>

=item L<Dancer::Plugin::DebugDump>

=item L<Dancer::Plugin::DebugToolbar>

=item L<Dancer::Plugin::DirectoryView>

=item L<Dancer::Plugin::Dispatcher>

=item L<Dancer::Plugin::ElasticSearch>

=item L<Dancer::Plugin::Email>

=item L<Dancer::Plugin::EncodeID>

=item L<Dancer::Plugin::EscapeHTML>

=item L<Dancer::Plugin::Facebook>

=item L<Dancer::Plugin::Fake::Response>

=item L<Dancer::Plugin::Feed>

=item L<Dancer::Plugin::FlashMessage>

=item L<Dancer::Plugin::FlashNote>

=item L<Dancer::Plugin::FormattedOutput>

=item L<Dancer::Plugin::FormValidator>

=item L<Dancer::Plugin::Hosts>

=item L<Dancer::Plugin::LibraryThing>

=item L<Dancer::Plugin::Memcached>

=item L<Dancer::Plugin::MemcachedFast>

=item L<Dancer::Plugin::MobileDevice>

=item L<Dancer::Plugin::Mongo>

=item L<Dancer::Plugin::Mongoose>

=item L<Dancer::Plugin::MPD>

=item L<Dancer::Plugin::Nitesi>

=item L<Dancer::Plugin::NYTProf>

=item L<Dancer::Plugin::ORMesque>

=item L<Dancer::Plugin::Params::Normalization>

=item L<Dancer::Plugin::Passphrase>

=item L<Dancer::Plugin::Preprocess::Sass>

=item L<Dancer::Plugin::Progress>

=item L<Dancer::Plugin::ProxyPath>

=item L<Dancer::Plugin::Redis>

=item L<Dancer::Plugin::REST>

=item L<Dancer::Plugin::Scoped> (temporarily disabled)

=item L<Dancer::Plugin::SimpleCRUD>

=item L<Dancer::Plugin::SiteMap>

=item L<Dancer::Plugin::SMS>

=item L<Dancer::Plugin::SporeDefinitionControl> (disabled)

=item L<Dancer::Plugin::Stomp>

=item L<Dancer::Plugin::TimeRequest>

=item L<Dancer::Plugin::Thumbnail>

=item L<Dancer::Plugin::ValidateTiny>

=item L<Dancer::Plugin::ValidationClass>

=item L<Dancer::Plugin::WebSocket>

=item L<Dancer::Plugin::XML::RSS>

=back

=head1 More Plack middlewares

=over 4

=item L<Dancer::Middleware::Rebase>

=item L<Dancer::Debug>

=back

=head1 AUTHOR

Sawyer X, C<xsawyerx AT cpan DOT org>

Alberto Simoes, C<ambs AT cpan DOT org>

=head1 BUGS

Please report any bugs or feature requests to C<bug-task-dancer at rt.cpan.org>,
or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Task-Dancer>.  I will be
notified, and then you'll automatically be notified of progress on your bug as I
make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Task::Dancer

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker (report bugs here)

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Task-Dancer>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/Task-Dancer>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/Task-Dancer>

=item * Search CPAN

L<http://search.cpan.org/dist/Task-Dancer/>

=back

=head1 ACKNOWLEDGEMENTS

L<Dancer> team.

=head1 LICENSE AND COPYRIGHT

Copyright 2010-2012 Sawyer X.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



inc/Module/Install/Compiler.pm
#line 1
package Module::Install::Compiler;

use strict;
use File::Basename        ();
use Module::Install::Base ();

use vars qw{$VERSION @ISA $ISCORE};
BEGIN {
	$VERSION = '0.93';
	@ISA     = 'Module::Install::Base';
	$ISCORE  = 1;
}

sub ppport {
	my $self = shift;
	if ( $self->is_admin ) {
		return $self->admin->ppport(@_);
	} else {
		# Fallback to just a check
		my $file = shift || 'ppport.h';
		unless ( -f $file ) {
			die "Packaging error, $file is missing";
		}
	}
}

sub cc_files {
	require Config;
	my $self = shift;
	$self->makemaker_args(
		OBJECT => join ' ', map { substr($_, 0, -2) . $Config::Config{_o} } @_
	);
}

sub cc_inc_paths {
	my $self = shift;
	$self->makemaker_args(
		INC => join ' ', map { "-I$_" } @_
	);
}

sub cc_lib_paths {
	my $self = shift;
	$self->makemaker_args(
		LIBS => join ' ', map { "-L$_" } @_
	);
}

sub cc_lib_links {
	my $self = shift;
	$self->makemaker_args(
		LIBS => join ' ', $self->makemaker_args->{LIBS}, map { "-l$_" } @_
	);
}

sub cc_optimize_flags {
	my $self = shift;
	$self->makemaker_args(
		OPTIMIZE => join ' ', @_
	);
}

1;

__END__

#line 123


Makefile.PL
use strict;
use warnings;

# Load the Module::Install bundled in ./inc/
use inc::Module::Install;

my %features = (
    templating => {
        namespace       => 'Template',

        Tiny            => { name => 'Template::Tiny',
                             min  => '0.02'                },
        HtmlTemplate    => { name => 'HTML::Template',
                             min  => '0.02'                },

        Alloy           => { name => 'Alloy'               },
        Tenjin          => { name => 'Tenjin'              },
        Haml            => { name => 'Text::Haml'          },
        Mason           => { name => 'HTML::Mason'         },
        Mason2          => { name => 'HTML::Mason2',
                             min  => '0.02'                },
        MicroTemplate   => { name => 'Text::MicroTemplate' },
        MojoTemplate    => { name => 'Mojo::Template wrapper for Dancer',
                             minperl => '5.010001' },
        Semantic        => { name => 'Semantic Template wrapper for Dancer',
                             min  => '0.08'                },
        TemplateFlute   => { name => 'Template::Flute wrapper for Dancer'    },
        TemplateSandbox => { name => 'Template::Sandbox'                     },
        Xslate          => { name => 'Text::Xslate wrapper for Dancer'       },
    },

    logging => {
        namespace  => 'Logger',

        ColorConsole => { name => 'A colorful console logger'                },
        LogHandler => { name => 'Log::Handler'                               },
        Log4perl   => { name => 'Log4Perl'                                   },
        Pipe       => { name => 'Pipe-based logging engine for Dancer'       },
        PSGI       => { name => 'PSGI'                                       },
        Spinner    => { name => 'Show a spinner in the console log messages' },
        Syslog     => { name => 'Sys::Syslog'                                },
    },

    serialization => {
        namespace => 'Serializer',

        UUEncode => { name => 'UU Encoding serializer' },
    },

    session => {
        namespace => 'Session',

        Cookie    => { name => 'Encrypted cookie-based'             },
        KiokuDB   => { name => 'KiokuDB Dancer session backend',
                       min  => '0.04'                               },
        Memcached => { name => 'Cache::Memcached'                   },
        MongoDB   => { name => 'MongoDB session backend for Dancer' },
        PSGI      => { name => 'Plack::Middleware::Session'         },
        Storable  => { name => 'Storable file-based'                },
        CHI       => { name => 'CHI-based session engine',
                       min => 0.001001  },
    },

    plugin => {
        namespace    => 'Plugin',

        Async           => { name => 'write async request handlers with Twiggy' },
        'Auth::Basic'   => { name => 'Basic HTTP authentication',
                             min => 0.011 },
        'Auth::Htpasswd'=> { name => 'Basic HTTP authentication with htpasswd files',
                             min => 0.011 },
        'Auth::RBAC'    => { name => 'Authentication, Security and Role-Based Access Control' },
        'Auth::Twitter' => { name => 'Authenticate with Twitter'                },
        Bcrypt          => { name => 'Bcrypt interface for Dancer'              },
        Browser         => { name => 'Browser detection'                        },
        'Cache::CHI'    => { name => 'Plugin to cache response content'         },
        'Captcha::SecurityImage' => {
                             name    => 'Generate and verify GD::SecurityImage captchas',
                             default => 0                          },
        Database        => { name => 'Database'                    },
        DBIC            => { name => 'DBIx::Class interface'       },
        DebugDump       => { name => 'dump objects to debug log'   },
        DebugToolbar    => { name => 'debugging toolbar'           },
        DirectoryView   => { name => 'allow directory browsing'    },
        Dispatcher   	=> { name => 'Controller Class Dispatching for Dancer' },
	ElasticSearch   => { name => 'Allow simple use of ElasticSearch' },
        Email           => { name => 'Email::Stuff integration'    },
        EncodeID        => { name => 'obfuscate IDs in UCLs'       },
        EscapeHTML      => { name => 'Escape HTML entities to avoid XSS vulnerabilities' },
        Facebook        => { name => 'Plugin linking Dancer with Facebook::Graph' },
        'Fake::Response' => { name => 'Fake request for inremental implementations' },
        Feed            => { name => 'Generate feed rss or atom'   },
        FlashMessage    => { name => 'Display temporary messages'  },
        FlashNote       => { name => 'Notification messages'       },
        FormattedOutput => { name => 'Provide output in a variety of formats' },
        FormValidator   => { name => 'Easy input validator'        },
        Hosts           => { name => 'Config trigger for virtual sites' },
        LibraryThing    => { name => 'Plugin for LibraryThing APIs' },
        Memcached       => { name => 'Cache response content'      },
        MemcachedFast   => { name => 'Cache things using Cache::Memcached::Fast' },
        MobileDevice    => { name => 'Dancer app mobile-aware',
                             min  => '0.03'                        },
        Mongo           => { name => 'MongoDB'                     },
	Mongoose        => { name => 'MongoDB <-> Moose ORM' 	   },
        MPD             => { name => 'Easy connection to MPD'      },
        Nitesi          => { name => 'Nitesi Shop Machine support' },
        NYTProf         => { name => 'Devel::NYTProf profiling',
                             min => '0.04'                         },
        ORMesque        => { name => 'Light ORM',
                             min => '1.112180'                     },
        Passphrase      => { name => 'Passphrases and Passwords as objects' },
        'Params::Normalization' => { name => 'query param normalization' },
        'Preprocess::Sass' => { name => 'Generate CSS files from Sass/SCSS files'},
        Progress        => { name => 'display a progress bar during long-running requests' },
        ProxyPath       => { name => 'Provides user-perspective paths' },
        Redis           => { name => 'Redis database connections'  },
        REST            => { name => 'RESTful apps',
                             min  => '0.05'                        },
        # Scoped        => { name => 'parameter variables scoping' }, => failing 0.01
        SimpleCRUD      => { name => 'very simple CRUD',
                             min  => '0.02'                        },
        SiteMap         => { name => 'Site map creation'           },
        SMS             => { name => 'Easy SMS sending'            },
        ## SporeDefinitionControl => { name => 'Validite route from a Spore' }, 0.09 is failing
        Stomp           => { name => 'Interact with Stomp message broker' },
        TimeRequests    => { name => 'Log how long each request took to process' },
        Thumbnail       => { name => 'Easy thumbnail creation',
                             default => 0      },
        ValidationClass => { name => 'Centralized Input Validation' },
        ValidateTiny    => { name => 'Validate::Tiny Dancer plugin' },
        WebSocket       => { name => 'Easily create WebSocket apps' },
        'XML::RSS'      => { name => 'XML::RSS to parse or create RSS feeds' },
    },

    # for now it's middlewares
    middleware => {
        namespace => 'Middleware',
        Rebase    => { name => 'Plack middleware for Dancer' },
    },

    other => {
        Debug => { name => 'Plack::Middleware::Debug' },
    },
);

name     'Task-Dancer';
all_from 'lib/Task/Dancer.pm';
license  'perl';
include  'Module::AutoInstall';
requires 'Dancer' => 0;

foreach my $type ( keys %features ) {
    my $namespace = delete $features{$type}{'namespace'};

    foreach my $semi_namespace ( keys %{ $features{$type} } ) {
        my $full_namespace = $namespace                               ?
                            "Dancer::${namespace}::${semi_namespace}" :
                            "Dancer::${semi_namespace}";

        if (exists($features{$type}{$semi_namespace}{minperl}) &&
            $features{$type}{$semi_namespace}{minperl} > $] ) {
            my $v = $features{$type}{$semi_namespace}{minperl};
            warn "**[ Skipping $full_namespace (requires Perl $v) ]**\n";
            next;
        }

        my ( $name, $default, $min ) =
            @{ $features{$type}{$semi_namespace} }{qw/ name default min /};

        # install by default unless otherwise specified
        $default = defined($default) ? 0 : 1;
        # no minimum version by default
        $min ||= 0;

        feature "$name $type support",
            -default => $default, $full_namespace => $min;
    }
}

auto_install;
WriteAll;

